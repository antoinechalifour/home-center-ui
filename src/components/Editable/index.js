import React, { Component } from 'react'
import PropTypes from 'prop-types'
import styled from 'styled-components'

export default class Editable extends Component {
  static propTypes = {
    children: PropTypes.string.isRequired,
    onChange: PropTypes.func.isRequired
  }

  state = {
    isEditionMode: false,
    input: ''
  }

  onKeyUp = e => {
    if (e.keyCode === 27) {
      // ESCAPE
      this.setState({
        isEditionMode: false,
        input: ''
      })
    } else if (e.keyCode === 13) {
      // ENTER
      this.props.onChange(this.state.input)
      this.setState({
        isEditionMode: false,
        input: ''
      })
    }
  }

  onChange = e => this.setState({ input: e.target.value })

  switchToEditionMode = () => {
    this.setState({
      isEditionMode: true,
      input: this.props.children
    })
  }

  render () {
    return this.state.isEditionMode
      ? <span>
        <Input
          value={this.state.input}
          onChange={this.onChange}
          onKeyUp={this.onKeyUp}
          />
      </span>
      : <span onClick={this.switchToEditionMode}>
        {this.props.children}
      </span>
  }
}

const Input = styled.input`
  display: block;
  width: 100%;
  box-sizing: border-box;
  font-size: inherit;
  font-family: inherit;
  border: 1px solid rgba(0, 0, 0, .14);
  outline: none;
  padding: 8px;
  border-radius: 4px;
`
