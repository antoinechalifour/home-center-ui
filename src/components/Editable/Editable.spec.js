import React from 'react'
import { shallow } from 'enzyme'
import Editable from '.'

describe('components/Editable', () => {
  it('Should render view mode', () => {
    const props = {
      onChange: jest.fn()
    }
    const wrapper = shallow(<Editable {...props}>this is editable</Editable>)

    expect(wrapper).toMatchSnapshot()
  })

  it('Should render edition mode', () => {
    const props = {
      onChange: jest.fn()
    }
    const wrapper = shallow(<Editable {...props}>this is editable</Editable>)

    wrapper.setState({
      isEditionMode: true,
      input: 'modified text'
    })

    expect(wrapper).toMatchSnapshot()
  })

  it('Should switch to edition mode on click', () => {
    const props = { onChange: jest.fn() }
    const wrapper = shallow(<Editable {...props}>test</Editable>)

    wrapper.simulate('click')

    expect(wrapper.state()).toEqual({
      isEditionMode: true,
      input: 'test'
    })
  })

  it('Should exit without calling onChange on key up 27', () => {
    const props = { onChange: jest.fn() }
    const wrapper = shallow(<Editable {...props}>test</Editable>)

    wrapper.setState({
      isEditionMode: true,
      input: 'updated text'
    })

    // Dirty hack to find styled component input
    const input = wrapper.findWhere(
      x => x.type().displayName === 'styled.input'
    )
    input.simulate('focus')
    input.simulate('keyup', { keyCode: 27 })

    expect(wrapper.state()).toEqual({
      isEditionMode: false,
      input: ''
    })
  })

  it('Should exit and call onChange on key up 13', () => {
    const props = { onChange: jest.fn() }
    const wrapper = shallow(<Editable {...props}>test</Editable>)

    wrapper.setState({
      isEditionMode: true,
      input: 'updated text'
    })

    // Dirty hack to find styled component input
    const input = wrapper.findWhere(
      x => x.type().displayName === 'styled.input'
    )
    input.simulate('focus')
    input.simulate('keyup', { keyCode: 13 })

    expect(wrapper.state()).toEqual({
      isEditionMode: false,
      input: ''
    })
    expect(props.onChange.mock.calls.length).toBe(1)
    expect(props.onChange.mock.calls[0][0]).toBe('updated text')
  })
})
