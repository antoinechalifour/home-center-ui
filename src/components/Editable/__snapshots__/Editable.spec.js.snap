// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`components/Editable Should render edition mode 1`] = `
.c0 {
  display: block;
  width: 100%;
  box-sizing: border-box;
  font-size: inherit;
  font-family: inherit;
  border: 1px solid rgba(0,0,0,.14);
  outline: none;
  padding: 8px;
  border-radius: 4px;
}

<span>
  <input
    className="c0"
    onChange={[Function]}
    onKeyUp={[Function]}
    value="modified text"
  />
</span>
`;

exports[`components/Editable Should render view mode 1`] = `
<span
  onClick={[Function]}
>
  this is editable
</span>
`;
